# **Section A – Output Prediction**

## ***Q1. Array Operations and Vectorization***

#### **Given Code**
```python
import numpy as np
import pandas as pd

arr = np.array([10, 20, 30])
s = pd.Series([10, 20, 30])

print(arr * 2)
print(s * 2)
```
##### **Code Explanation:** 
Here, we created a NumPy array and a Pandas Series with the same values. We multiplied both by 2 to see how the values change using vectorized operations.

### ***1. Predict the output of both statements.***


**Output:**
```text
[20 40 60]
0       20
1       40
2       60
dtype: int64
```

The NumPy array multiplies every value by 2. The Pandas Series also multiplies every value by 2, but it displays the index along with the values.


### ***2. Identify one similarity and one difference between the two data structures.***


##### **Similarity:**

Both NumPy arrays and Pandas Series allow us to perform operations on all values at once without using a for loop.

##### **Difference:**

A NumPy array mainly stores values, while a Pandas Series also has an index for each value.

## ***Q2. Boolean Filtering***

#### **Given Code**
```python
import pandas as pd

df = pd.DataFrame({
    'Student': ['A', 'B', 'C', 'D'],
    'Marks': [45, 80, 65, 90]
})

result = df[df['Marks'] >= 65]

print(result[['Student', 'Marks']])
```
##### **Code Explanation:** 
Here, we created a DataFrame with student names and marks. We used a condition to select students who scored 65 or more and displayed only their names and marks.

### ***1. Predict the rows displayed.***

**Rows displayed:**

The condition is Marks >= 65, so students B, C, and D satisfy the condition.
```text
  Student  Marks
1       B     80
2       C     65
3       D     90
```

### ***2. How many rows and columns will the resulting DataFrame contain?***

**Number of rows and columns:**

Rows = 3
Columns = 2

So, the resulting DataFrame has a shape of (3, 2).

The filtering keeps only the students whose marks are 65 or above.

## ***Q3. Descriptive Statistics***

#### **Given Code**
```python
import pandas as pd

df = pd.DataFrame({
    'Marks': [10, 20, 30, 40, 50]
})

print(df['Marks'].mean())
print(df['Marks'].max())
print(df['Marks'].median())
```

##### **Code Explanation:**
Here, we created a DataFrame with marks and used `mean()`, `max()`, and `median()` to find the average, highest, and middle values.

### ***1. Predict all three outputs.***

**Outputs:**
```text
30.0
50
30.0
```
Mean = 30.0

Maximum = 50

Median = 30.0

### ***2. Explain how the median differs from the mean.***

**Difference between mean and median:**

Mean is calculated by adding all the values and dividing by the number of values.

Median is the middle value after arranging the values in order.

Here, both mean and median are 30.0.

## ***Q4. Group-wise Analysis***

#### **Given Code**
```python
import pandas as pd

df = pd.DataFrame({
    'Department': ['IT', 'HR', 'IT', 'HR'],
    'Salary': [40000, 30000, 60000, 50000]
})

print(df.groupby('Department')['Salary'].mean())
```

##### **Code Explanation:** 
Here, we grouped the employees based on their department and used `mean()` to find the average salary for each department.

### ***1. Predict the output.***

**Output:**
```text
Department
HR    40000.0
IT    50000.0
Name: Salary, dtype: float64
```

### ***2. Explain how groupby() calculates the department-wise average.***

**Explanation:**

`groupby()` first separates the data based on the department.

For HR:

(30000 + 50000) / 2 = 40000

For IT:

(40000 + 60000) / 2 = 50000

So, it calculates the average salary separately for each department.

## ***Q5. Pandas Correlation – Positive***

#### **Given Code**
```python
import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))
```


##### **Code Explanation:** 
Here, we created data for study hours and marks. We used `corr()` to check how strongly the study hours and marks are related.


### ***1. Predict the correlation coefficient.***

**Correlation coefficient:**
```text
1.0
```


### ***2. Interpret the direction and strength of the relationship.***

**Interpretation:**

The correlation is positive and very strong. It is 1.0, which means that as the study hours increase, the marks also increase in a perfectly positive relationship.

In this data, every increase in hours is followed by a proportional increase in marks.

## ***Q6. Pandas Correlation – Negative***

#### **Given Code**
```python
import pandas as pd

df = pd.DataFrame({
    'Price': [10, 20, 30, 40],
    'Demand': [100, 80, 60, 40]
})

print(df['Price'].corr(df['Demand']))
```

##### **Code Explanation:** Here, we created data for price and demand. We used `corr()` to check the relationship between them and see whether they increase or decrease together.

### ***1. Predict the correlation coefficient.***

**Correlation coefficient:**
```text
-1.0
```

### ***2. Explain what the sign of the coefficient indicates about the relationship.***

**Explanation:**

The negative sign shows that Price and Demand have an opposite relationship.

Here, when the price increases, the demand decreases. The value -1.0 means the relationship is perfectly negative for these values.

# **Section B – Code Correction**


### ***Q1. JSON and CSV***

#### **Given Code:**
```python
import pandas as pd

df = pd.read_csv("students.csv", header=None)
print(df)
```

### **Issue:**

The CSV file already contains a header row. Using `header=None` tells Pandas that the file has no header, so the first row can be treated as normal data instead of column names.

### **Corrected Code:**
```python
import pandas as pd

df = pd.read_csv("students.csv")
print(df)

```

### **Explanation:**

I removed `header=None` so that Pandas can use the first row of the CSV file as the column names.

### ***Q2. Data Analysis***

#### **Given Code:**
```python
import pandas as pd

df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary').head(5))
```

### **Issue:**

The code sorts the salary values in ascending order, so `head(5)` gives the five lowest salaries. The objective is to display the five highest salaries.

### **Corrected Code:**

```python
import pandas as pd

df = pd.DataFrame({
    'Employee': ['A', 'B', 'C', 'D', 'E', 'F'],
    'Salary': [30000, 50000, 25000, 60000, 45000, 70000]
})

print(df.sort_values('Salary', ascending=False).head(5))
```

### **Explanation:**

I used `ascending=False` to sort salaries from highest to lowest and then selected the first five rows.

### ***Q3. Cleaning Data***

#### **Given Code:**
```python
import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})

df['Marks'].fillna(df['Marks'].mean())
print(df)
```

### **Issue:**

`fillna()` creates the updated values, but the result is not assigned back to the Marks column. Therefore, the original DataFrame is not changed.

### **Corrected Code:**
```python
import pandas as pd

df = pd.DataFrame({'Marks': [70, None, 90]})

df['Marks'] = df['Marks'].fillna(df['Marks'].mean())
print(df)
```

### **Explanation:**

I assigned the result of `fillna()` back to the Marks column so that the missing value is replaced with the column mean.

### ***Q4. Correlation***

#### **Given Code:**
```python
import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df.corr('Hours', 'Marks'))
```

### **Issue:**

`corr()` is being used incorrectly. To calculate the correlation between two specific columns, we can select both columns as Series and use `corr()`.

### **Corrected Code:**
```python
import pandas as pd

df = pd.DataFrame({
    'Hours': [1, 2, 3, 4],
    'Marks': [20, 40, 60, 80]
})

print(df['Hours'].corr(df['Marks']))
```

### **Explanation:**

I selected the Hours and Marks columns separately and used `corr()` to calculate their correlation coefficient.

# **Section C – Edge Cases**

### ***Q1. NumPy Broadcasting***

**Given Code**
```python
import numpy as np

a = np.array([[10], [20], [30]])
b = np.array([[1, 2, 3]])

result = a + b

print(result)
print(result.shape)
```
### ***1. Predict the output***
```test
[[11 12 13]
 [21 22 23]
 [31 32 33]]
(3, 3)
```
### ***2. Identify the shapes of a and b***
```text 
Shape of a = (3, 1)
Shape of b = (1, 3)
```

### ***3. Explain how broadcasting enables the addition***

Broadcasting allows NumPy to expand the smaller dimensions so that the two arrays can be added together. Here, a is treated like 3 rows and b is treated like 3 columns, so the final result has a shape of (3, 3).

### ***4. What happens if b = np.array([1, 2])?***

The addition will give a ValueError because the shapes (3, 1) and (2,) are not compatible for broadcasting. The dimensions 3 and 2 do not match.

##### **Code Explanation:** 
Here, broadcasting allows the (3, 1) and (1, 3) arrays to be added without manually changing their shapes.

### ***Q2. Messy CSV Data***

**Given Dataset:**

ID,Name,Date,Marks
101,"Asha, R",2026-09-01,80
102,Ravi,01/09/2026,
103,Neha,2026-09-03,NA

**Given Code**
```python
import pandas as pd

df = pd.read_csv("students.csv")

print(df)
print(df.dtypes)
```

### ***1. How is the quoted comma in "Asha, R" handled?***

The comma inside "Asha, R" is treated as part of the name because the name is enclosed in double quotes. It will be read as one field instead of creating an extra column.

### ***2. What are the potential issues with Date and Marks?***

**Date:**
```text
The Date column contains different date formats:

2026-09-01
01/09/2026
2026-09-03
```
These should be converted into a consistent date-time format.

**Marks:**
The Marks column contains a blank value and NA, which need to be handled as missing values.

### ***3. Rewrite the code to handle missing marks and convert the date***
```python
import pandas as pd

df = pd.read_csv("students.csv", na_values=["NA"])

df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")
df["Date"] = pd.to_datetime(df["Date"], errors="coerce")

print(df)
print(df.dtypes)
```
### ***4. Why should "NA" be investigated before treating it as missing?***

NA may mean that the mark was not entered, the student was absent, or the data was unavailable. We should understand its meaning before deciding how to handle it.

##### **Code Explanation:**

Here, `read_csv()` reads the file, `to_numeric()` converts marks into numeric values, and `to_datetime()` converts the date column into a date-time format.

## ***Q3. Salary Outlier***

**Given Code:**
```python
import pandas as pd

df = pd.DataFrame({
    "Employee": ["A", "B", "C", "D"],
    "Salary": [25000, 30000, 35000, 10000000]
})

print("Mean:", df["Salary"].mean())
print("Median:", df["Salary"].median())
```

### ***1. Predict the mean and median salary***

**Mean:**

Mean: 2540000.0

**Median:**

Median: 32500.0

### ***2. Effect of the extreme salary on the mean***

The salary 10000000 is much higher than the other salaries. It increases the mean significantly, making the mean much higher than the salaries of most employees.

### ***3. Which statistic better represents a typical employee's salary?***

The median better represents the typical salary in this dataset because it is less affected by the extreme value.

The median salary is 32500.0, which is closer to the salaries of employees A, B, and C.

### ***4. Identify the employee with the highest salary***
```python
highest_salary = df.loc[df["Salary"].idxmax()]

print(highest_salary)
```

**Output:**
```text
Employee           D
Salary       10000000
Name: 3, dtype: object
```

The record should be investigated before removing it because the value may be a genuine high salary rather than an error. Removing it without checking could lead to loss of valid data.

##### **Code Explanation**

Here, we compare the mean and median to understand the effect of the extreme salary and use `idxmax()` to identify the employee with the highest salary.

# ***Section D – Scenario-Based Questions***

## ***Q1. Student Assessment Data Cleaning***

#### **Question**

A student assessment dataset contains missing marks, duplicate student records, inconsistent department names such as IT, it, and I.T., and marks stored as text. Develop a cleaning sequence to standardize the data, handle missing values, remove duplicates, and convert marks into a suitable numeric type. State one risk of deleting missing values without investigating why they are missing.

**Dataset:** `student_assessment_dirty.csv`

## ***1. Load and Inspect the Dataset***


In [1]:
import pandas as pd

df = pd.read_csv("student_assessment_dirty.csv", keep_default_na=False)

print(df.head())
print(df.info())
print(df.isnull().sum())

#Code Explanation: Here, I loaded the CSV file and used head(), info(), and isnull().sum() to understand the dataset and check for missing values.

  Student_ID Student_Name Department Assessment   Marks
0       S101         Asha         IT     Python      85
1       S102         Ravi         it     Python      78
2       S103         Neha       I.T.     Python      92
3       S104        Karan         HR     Python        
4       S105        Meera         HR     Python  Absent
<class 'pandas.DataFrame'>
RangeIndex: 15 entries, 0 to 14
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   Student_ID    15 non-null     str  
 1   Student_Name  15 non-null     str  
 2   Department    15 non-null     str  
 3   Assessment    15 non-null     str  
 4   Marks         15 non-null     str  
dtypes: str(5)
memory usage: 732.0 bytes
None
Student_ID      0
Student_Name    0
Department      0
Assessment      0
Marks           0
dtype: int64


## ***2. Standardize Department Names***

In [2]:
df["Department"] = df["Department"].str.strip().str.lower()

df["Department"] = df["Department"].replace({
    "it": "IT",
    "i.t.": "IT",
    "hr": "HR",
    "finance": "Finance"
})

print(df["Department"])


#Code Explanation: Here, I standardized department names so that different forms such as IT, it, and I.T. are represented consistently.

0          IT
1          IT
2          IT
3          HR
4          HR
5     Finance
6     Finance
7          IT
8          IT
9          IT
10         HR
11    Finance
12         IT
13         HR
14    Finance
Name: Department, dtype: str


## ***3. Remove Exact Duplicate Recordsv***

In [3]:
before = len(df)

df = df.drop_duplicates()

removed = before - len(df)

print("Records removed:", removed)

#Code Explanation: Here, I used drop_duplicates() to remove the exact duplicate record of student S102.

Records removed: 1


## ***4. Convert Marks into Numeric Format***

The dataset contains different types of values in the Marks column:

Blank → mark not entered

Absent → student was absent

NA → mark not entered

88 marks → should become 88

Numeric values → converted to numbers

In [4]:
import numpy as np

conditions = [
    df["Marks"].eq("Absent"),
    df["Marks"].isin(["", "NA"])
]

choices = ["Absent", "Not Entered"]

df["Mark_Status"] = np.select(
    conditions,
    choices,
    default="Entered"
)

df["Marks"] = df["Marks"].str.extract(
    r"(\d+(?:\.\d+)?)",
    expand=False
)

df["Marks"] = pd.to_numeric(df["Marks"], errors="coerce")

print(df)

#Code Explanation: Here, I separated the status of missing marks and converted values such as 88 marks into numeric values.

   Student_ID Student_Name Department Assessment  Marks  Mark_Status
0        S101         Asha         IT     Python   85.0      Entered
1        S102         Ravi         IT     Python   78.0      Entered
2        S103         Neha         IT     Python   92.0      Entered
3        S104        Karan         HR     Python    NaN  Not Entered
4        S105        Meera         HR     Python    NaN       Absent
5        S106        Arjun    Finance     Python    0.0      Entered
6        S107        Priya    Finance     Python   67.0      Entered
7        S108        Rahul         IT     Python   75.0      Entered
9        S109        Sneha         IT     Python    NaN  Not Entered
10       S110        Varun         HR     Python   88.0      Entered
11       S111         Diya    Finance     Python   91.0      Entered
12       S112         Aman         IT     Python    NaN  Not Entered
13       S113        Nisha         HR     Python   64.0      Entered
14       S114        Kabir    Fina

### ***Important Result***

#### **For example:**
```text
88 marks → 88
Absent   → NaN
NA       → NaN
blank    → NaN
```
The Mark_Status column helps us distinguish Absent from marks that were not entered.

## ***5. Investigate Missing Marks***

In [5]:
print(df[df["Marks"].isna()][
    ["Student_ID", "Student_Name", "Marks", "Mark_Status"]
])


   Student_ID Student_Name  Marks  Mark_Status
3        S104        Karan    NaN  Not Entered
4        S105        Meera    NaN       Absent
9        S109        Sneha    NaN  Not Entered
12       S112         Aman    NaN  Not Entered


### ***Explanation***

The missing marks should not all be treated in the same way.

Meera is **Absent**.
Karan, Sneha, and Aman have marks that were not entered.
Therefore, these values should be investigated before deleting them.

### ***Risk of deleting missing values***

Deleting missing values without investigating them may remove useful student records and can lead to incorrect analysis.

## ***6. Identify Marks Outside the Valid Range***

The valid mark range is 0 to 100.

In [6]:
invalid_marks = df[
    df["Marks"].notna() &
    ~df["Marks"].between(0, 100)
]

print(invalid_marks)


   Student_ID Student_Name Department Assessment  Marks Mark_Status
14       S114        Kabir    Finance     Python  105.0     Entered


### ***Explanation***

Kabir has a mark of 105, which is outside the valid range of 0–100. This value should be investigated before changing or removing it because it could be a data-entry error.

## ***7. Display the Cleaned DataFrame***

In [7]:
print(df)

#Code Explanation: Here, the cleaned DataFrame shows standardized departments, numeric marks, missing-mark status, and the invalid mark that needs investigation.

   Student_ID Student_Name Department Assessment  Marks  Mark_Status
0        S101         Asha         IT     Python   85.0      Entered
1        S102         Ravi         IT     Python   78.0      Entered
2        S103         Neha         IT     Python   92.0      Entered
3        S104        Karan         HR     Python    NaN  Not Entered
4        S105        Meera         HR     Python    NaN       Absent
5        S106        Arjun    Finance     Python    0.0      Entered
6        S107        Priya    Finance     Python   67.0      Entered
7        S108        Rahul         IT     Python   75.0      Entered
9        S109        Sneha         IT     Python    NaN  Not Entered
10       S110        Varun         HR     Python   88.0      Entered
11       S111         Diya    Finance     Python   91.0      Entered
12       S112         Aman         IT     Python    NaN  Not Entered
13       S113        Nisha         HR     Python   64.0      Entered
14       S114        Kabir    Fina

## ***8. Final Report***

In [8]:
print("Records removed:", removed)
print("Missing marks remaining:", df["Marks"].isna().sum())
print("Invalid marks identified:", len(invalid_marks))


Records removed: 1
Missing marks remaining: 4
Invalid marks identified: 1


### ***Final Explanation***

The dataset was cleaned by standardizing department names, removing the duplicate record, converting marks into numeric values, identifying missing marks, and detecting the invalid mark 105. Missing and invalid values were investigated instead of being blindly deleted.